### PR TITLE
Datastore tutorial

### PR DESCRIPTION
**Pull request by irisliu77**

### CODE REVIEW

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 1]
```diff
@@ -21,29 +21,51 @@
...
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    String json = convertToJson(comments);
+    Query query = new Query("Comment").addSort("timestamp", SortDirection.DESCENDING);
+
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    PreparedQuery res = datastore.prepare(query);
+
+    List<Entity> comments = new ArrayList<>();
+    for(Entity entity : res.asIterable()) {
+        String content = (String) entity.getProperty("content");
```
Comment (anbenson) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:45]: So the code here feels a bit strange - you have an Entity object, deconstruct it into its parts, then put it back together again to add to the ArrayList. Why not add it right away?

But stepping back for a second - the code on line 54 turns the Entity object (which you don't know the internals of, it being from a 3rd party library) into JSON instead of turning a format you have control over (such as the Comment class you presumably are using in your other PR). This makes it so that your JavaScript below is forced to examine internal properties of the object which could change if the Entity library decided to change, which is brittle.

In summary, I would advise against serializing the Entity object directly to JSON. You can use the Comment object now, or since the only thing your JS uses is the content property, you can just serialize an ArrayList<String> of the content properties.

Let me know if this is confusing.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 2]
```diff
@@ -21,29 +21,51 @@
...
   }
 
   @Override
   public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
       String text = request.getParameter("comment-input");
-      comments.add(text);
+      long timestamp = System.currentTimeMillis();
+
+      Entity comment = new Entity("Comment");
+      comment.setProperty("content", text);
```
Comment (anbenson) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:63]: These string names ("content", "timestamp") are a bit fragile since they're constants that are used in more than one place. Can you make them `private static final String` constants instead? This protects against typos and partial changes to the names.

File: portfolio/src/main/webapp/script.js [hunk 3]
```diff
@@ -57,16 +57,23 @@ const getComments = function() {
     fetch('/data').then(response => response.json()).then((comments) => {
         const commentsContainer = document.getElementById('comments-container');
         commentsContainer.innerHTML = '';
-        
-        for(let i = 0; i < comments.length; i++) {
-            commentsContainer.appendChild(createListElement('Comment: ' + comments[i]));
-        }
+        comments.forEach((Comment) => {
```
Comment (anbenson) [inline portfolio/src/main/webapp/script.js:60]: `Comment` should not be capitalized since it is a variable name.

I recommend against the Array.prototype.forEach pattern - it's inverted, so it's harder to read at first glance than a for-of loop (I know I keep bringing for-of loops up, but it was created to fix all the problems with forEach, for-in, and for (let i = 0; i < ...)), and it's harder to debug since it takes in a function as an argument so you can't step directly into it in the JS debugger.

File: portfolio/src/main/webapp/script.js [hunk 4]
```diff
@@ -57,16 +57,23 @@ const getComments = function() {
...
 /** Creates an <li> element containing text. */
-const createListElement = function(text) {
-    const liElement = document.createElement('li');
-    liElement.innerText = text;
-    return liElement;
+function createCommentElement(comment) {
+    const commentElement = document.createElement('li');
+    commentElement.className = 'comment';
+    commentElement.innerText = comment.key.kind;
+
```
Comment (anbenson) [inline portfolio/src/main/webapp/script.js:71]: Remove extra whitespace

File: portfolio/src/main/webapp/script.js [hunk 5]
```diff
@@ -57,16 +57,23 @@ const getComments = function() {
...
+    return commentElement;
+}
+
+const handleSubmitButton = function() {
```
Comment (anbenson) [inline portfolio/src/main/webapp/script.js:75]: Clicking the submit button submits the form, which triggers a reload of the page, so this JS hardly has a chance to execute (and even if it did, it wouldn't get the new comment since that would require the reload to finish).

**Comment by irisliu77** [comment]
ready for re-review!

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 1 above)
Comment (joanchenx) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:48]: You're not using the `content` and `timestamp` constants you added in line 37-38. And you end up using the same variable name inside your for-loop.

Perhaps, rename those constants to `contentParameter` and `timestampParameter` to avoid confusion. And then use those parameters when grabbing the content/timestamp:
`String content = (String) entity.getProperty(contentParameter);`

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 6]
```diff
@@ -21,29 +21,54 @@
...
 
   @Override
   public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
       String text = request.getParameter("comment-input");
-      comments.add(text);
+      long time = System.currentTimeMillis();
+
+      Entity comment = new Entity("Comment");
+      comment.setProperty(content, text);
+      comment.setProperty(timestamp,time);
```
Comment (joanchenx) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:67]: Nit: add a space after the comma

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 7]
```diff
@@ -21,29 +21,54 @@
...
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Query.SortDirection;
 
 /** Servlet that returns some example content. TODO: modify this file to handle comments data */
 @WebServlet("/data")
 public class DataServlet extends HttpServlet {
-  private final ArrayList<String> comments = new ArrayList<String>();
-  
+    private static final String contentParameter = "content";
+    private static final String timestampParameter = "timestamp";
```
Comment (joanchenx) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:38]: Nit: add a new line after this line.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 8]
```diff
@@ -21,29 +21,54 @@
...
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    String json = convertToJson(comments);
+    Query query = new Query("Comment").addSort("timestamp", SortDirection.DESCENDING);
```
Comment (joanchenx) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:41]: You can use `timestampParameter` here instead of the hardcoded `"timestamp"`

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 7 above)
Comment (joanchenx) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:38]: Whoops, sorry I gave you incorrect instructions earlier. For constants, name them snake case, ie. `CONTENT_PARAMETER` instead.

File: portfolio/src/main/webapp/script.js [hunk 9]
```diff
@@ -56,17 +56,17 @@ const handleClickProject = function() {
...
     fetch('/data').then(response => response.json()).then((comments) => {
         const commentsContainer = document.getElementById('comments-container');
-        commentsContainer.innerHTML = '';
-        
+        commentsContainer.innerHTML = ''; 
         for(let comment of comments) {
-            commentsContainer.appendChild(createListElement('Comment: ' + comment));
-        }
+            commentsContainer.appendChild(createCommentElement(comment));
+        };
```
Comment (anbenson) [inline portfolio/src/main/webapp/script.js:62]: No semicolon after a for loop.

File: portfolio/src/main/webapp/script.js [hunk 10]
```diff
@@ -56,17 +56,17 @@ const handleClickProject = function() {
 const getComments = function() {
     fetch('/data').then(response => response.json()).then((comments) => {
         const commentsContainer = document.getElementById('comments-container');
-        commentsContainer.innerHTML = '';
-        
+        commentsContainer.innerHTML = ''; 
         for(let comment of comments) {
```
Comment (anbenson) [inline portfolio/src/main/webapp/script.js:60]: Space between for and (

File: portfolio/src/main/webapp/style.css [hunk 11]
```diff
@@ -91,3 +91,7 @@
     margin-top: 20px;
     margin-bottom: 20px;
 }
+
+.comment{
```
Comment (anbenson) [inline portfolio/src/main/webapp/style.css:95]: Space between comment and {
